### PR TITLE
Media Library: add keyboard navigation

### LINKED ISSUE #1305: Media Library - Keyboard Navigation
## Feature Description

As an author I want to be able to navigate the media library via keyboard

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

1. Implement keyboard navigation. 
2. Keyboard should trigger infinite scroll loading for next page.

### PR DESCRIPTION
**Pull request by diegovar**
## Summary

This PR adds keyboard navigation to the media gallery (both Upload and Media3P tabs) by implementing the "roving tabindex" aria recommendation.
See https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex

## Relevant Technical Choices

Decided not to use aria-activedescendant as it's not clear at all what the browser support is.

## To-do

I believe there's an issue with keyboard control in the canvas where even though the focus is in the gallery, using the kyeboard results in elements in the canvas moving.

## User-facing changes

When tabbing through the editor and reaching the gallery, the first element in the gallery is outlined. At that point you can use Up/Down/Left/Right/Home/End to navigate the gallery. Tabbing again exits the gallery onto the next tabindex (in the canvas I believe).

If a media element is selected and the user presses Enter, the media element is added to the canvas (equivalent to clicking the element).

## Testing Instructions

Enable the rowBasedGallery and media3pTab flags.

See user-facing changes.

---

Fixes #1305

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A point in 2D space.
```
Comment (cvolzke4) [inline assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js:18]: This comment doesn't apply to any code.

**Comment by cvolzke4** [comment]
I noticed a couple of things:
- When pressing page-down, the image selection doesn't update with it. So if I move down 3 pages, then use the arrows to pick an image, the scroll window moves back to the top.
- When holding your finger on the down arrow, the 'Loading' pill doesn't always appear once you reach the bottom of the page.

File: assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js (hunk 1 above)
Comment (diegovar) [reply assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js:18]: It's a typedef, see getResourceFromMedia3p for another example.

**Comment by diegovar** [comment]
> I noticed a couple of things:
> 
> * When pressing page-down, the image selection doesn't update with it. So if I move down 3 pages, then use the arrows to pick an image, the scroll window moves back to the top.
> * When holding your finger on the down arrow, the 'Loading' pill doesn't always appear once you reach the bottom of the page.

Next/Prev page are not currently handled. It would be up to us to determine what the "page" size should be, WDYT?

The second looks like a bug (which shouldn't happen?), I'll investigate. Might be a consequence for no longer using an intersection observer

**Comment by diegovar** [comment]
I've added karma tests

File: assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js (hunk 1 above)
Comment (cvolzke4) [reply assets/src/edit-story/components/library/panes/media/common/useRovingTabIndex.js:18]: Ahh, that's quite nice.

**Comment by spacedmonkey** [comment]
This is really good! 

Two things to note. 

1. When I am focus on an image, how do I media options dropdown. Could I press space or something to show these menu?
Screenshot 2020-07-21 at 13 24 40
2. Can we use the [speak](https://github.com/WordPress/gutenberg/blob/00aa937ac873e8e697f4354efce11bcbf3c74a70/packages/a11y/src/index.js#L53-L69) api, to announce to screen readers the directions to use the keyboard navigation.

**Comment by diegovar** [comment]
> I noticed a couple of things:
> 
> * When pressing page-down, the image selection doesn't update with it. So if I move down 3 pages, then use the arrows to pick an image, the scroll window moves back to the top.
> * When holding your finger on the down arrow, the 'Loading' pill doesn't always appear once you reach the bottom of the page.

I've added handling for pagedown and pageup.

I wasn't able to replicate the issue with the Loading pill not appearing, are you sure you had more elements after that?

**Comment by diegovar** [comment]
> This is really good!
> 
> Two things to note.
> 
> 1. When I am focus on an image, how do I media options dropdown. Could I press space or something to show these menu?
>    Screenshot 2020-07-21 at 13 24 40
> 2. Can we use the [speak](https://github.com/WordPress/gutenberg/blob/00aa937ac873e8e697f4354efce11bcbf3c74a70/packages/a11y/src/index.js#L53-L69) api, to announce to screen readers the directions to use the keyboard navigation.

I've added support to open the media dropdown when the users presses Space, the dropdown first item gets selected automatically :)

Wrt to the `speak` API, I'd rather understand the best practices about announcing this sort of thing better before adding that, so I'd tackle the investigation in a separate ticket.

**Comment by cvolzke4** [comment]
> > I noticed a couple of things:
> > 
> > * When pressing page-down, the image selection doesn't update with it. So if I move down 3 pages, then use the arrows to pick an image, the scroll window moves back to the top.
> > * When holding your finger on the down arrow, the 'Loading' pill doesn't always appear once you reach the bottom of the page.
> 
> I've added handling for pagedown and pageup.
> 
> I wasn't able to replicate the issue with the Loading pill not appearing, are you sure you had more elements after that?

To reproduce it, navigate to the Media3pPane and hold down the down arrow:
https://share.getcloudapp.com/E0uzkJWP

**Review by cvolzke4 — APPROVED** [review]

**Comment by cvolzke4** [comment]
Looks great

**Pull request merged**